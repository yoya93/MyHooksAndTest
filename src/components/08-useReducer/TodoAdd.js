import React from "react";
import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({ handleAdd }) => {
  const [value, handleInputChange, reset] = useForm({
    description: "",
  });

  const { description } = value;

  const handleSubmitAdd = (e) => {
    e.preventDefault();

    if (description.trim().length < 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };

    handleAdd(newTodo);

    reset();
  };

  return (
    <>
      <h3>Agregar Todo</h3>
      <hr />

      <form onSubmit={handleSubmitAdd}>
        <input
          className="form-control"
          value={description}
          name="description"
          type="text"
          placeholder="Aprender ....."
          autoComplete="off"
          onChange={handleInputChange}
        ></input>

        <button
          type="submit"
          className="btn btn-outline-primary btn-block mt-1"
        >
          Agregar
        </button>
      </form>
    </>
  );
};
