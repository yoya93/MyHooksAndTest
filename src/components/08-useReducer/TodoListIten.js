import React from "react";

export const TodoListIten = ({ i, todo, handleThrough, handleDeleteIten }) => {
  return (
    <>
      <li key={todo.id} className="list-group-iten mb-2">
        <p
          onClick={() => {
            handleThrough(todo.id);
          }}
          className={`${todo.done && "complete"}`}
        >
          {i + 1}-{todo.desc}
        </p>
        <button
          onClick={() => {
            handleDeleteIten(todo.id);
          }}
          className="btn btn-danger"
        >
          Borrar
        </button>
      </li>
    </>
  );
};
