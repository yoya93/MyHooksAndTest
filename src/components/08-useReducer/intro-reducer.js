const initialState = [
  {
    id: 1,
    todo: "Comprar leche",
    done: false,
  },
];

const todoReducer = (state = initialState, AddTodoAction) => {
  if (AddTodoAction?.type === "agregar") {
    return [...state, AddTodoAction.pyload];
  }
  return state;
};

let todos = todoReducer();

const newTod = {
  id: 2,
  todo: "Comprar Pan",
  done: false,
};

const AddTodoAction = {
  type: "agregar",
  pyload: newTod,
};

todos = todoReducer(todos, AddTodoAction);

console.log(todos);
