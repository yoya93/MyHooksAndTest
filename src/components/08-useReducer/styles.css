body {
  padding: 50px;
}

p {
  cursor: pointer;
  margin: 0px;
}

.list-group-iten {
  align-items: center;
  display: flex;
  justify-content: space-between;
}

.complete {
  text-decoration: line-through;
}
