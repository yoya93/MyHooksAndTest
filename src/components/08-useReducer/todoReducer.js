export const todoReducer = (state = [], action) => {
  switch (action.type) {
    case "add":
      return [...state, action.payload];
    case "delete":
      return state.filter((todo) => todo.id !== action.payload); //filter regresa un arreglo con todos los itens que cumplan esa condicion
    case "through":
      return state.map((todo) =>
        todo.id === action.payload ? { ...todo, done: !todo.done } : todo
      );

    default:
      return state;
  }
};

// case "through":
//   return state.map((todo) => {
//     if (todo.id === action.payload) {
//       // busco el elemento del array ke se kiere cambiar
//       return { ...todo, done: !todo.done }; //retorno el objeto con la propiedad done cambiada
//     } else {
//       return todo;
//     }
//   });
