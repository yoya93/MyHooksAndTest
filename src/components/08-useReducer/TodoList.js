import React from "react";
import { TodoListIten } from "./TodoListIten";

export const TodoList = ({ todos, handleThrough, handleDeleteIten }) => {
  return (
    <>
      <ul className="list-group list-group-flush">
        {todos.map((todo, i) => (
          <TodoListIten
            key={todo.id}
            todo={todo}
            i={i}
            handleDeleteIten={handleDeleteIten}
            handleThrough={handleThrough}
          />
        ))}
      </ul>
    </>
  );
};
