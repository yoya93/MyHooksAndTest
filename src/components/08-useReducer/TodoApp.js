import React, { useReducer, useEffect } from "react";
import { todoReducer } from "./todoReducer";

import { TodoAdd } from "./TodoAdd";
import { TodoList } from "./TodoList";

import "./styles.css";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
  //[
  //     {
  //       id: new Date().getTime(),
  //       desc: "Aprender React",
  //       done: false,
  //     },
  //   ];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const handleDeleteIten = (id) => {
    const action = {
      type: "delete",
      payload: id,
    };

    dispatch(action);
  };

  const handleThrough = (id) => {
    dispatch({
      type: "through",
      payload: id,
    });
  };

  const handleAdd = (newTodo) => {
    dispatch({
      type: "add",
      payload: newTodo,
    });
  };

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  return (
    <>
      <h1>TodoApp({todos.length})</h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            handleDeleteIten={handleDeleteIten}
            handleThrough={handleThrough}
            todos={todos}
          />
        </div>

        <div className="col-5">
          <TodoAdd handleAdd={handleAdd} />
        </div>
      </div>
    </>
  );
};
