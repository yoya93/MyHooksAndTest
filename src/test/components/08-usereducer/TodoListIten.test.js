import React from "react";
import { TodoListIten } from "../../../components/08-useReducer/TodoListIten";
import { todosDemo } from "../../fixtures";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";

describe("pruebas en el TodoListIten", () => {
  const handleThrough = jest.fn();
  const handleDeleteIten = jest.fn();
  test("debe mostrarse correctamente", () => {
    const wrapper = shallow(
      <TodoListIten
        i={0}
        todo={todosDemo[0]}
        handleThrough={handleThrough}
        handleDeleteIten={handleDeleteIten}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("debe llamar loa funcion handleThrough ", () => {
    const wrapper = shallow(
      <TodoListIten
        i={0}
        todo={todosDemo[0]}
        handleThrough={handleThrough}
        handleDeleteIten={handleDeleteIten}
      />
    );

    wrapper.find("p").simulate("click");

    expect(handleThrough).toHaveBeenCalledWith(todosDemo[0].id);
  });

  test("debe llamar la funcion handleDeleteIten ", () => {
    const wrapper = shallow(
      <TodoListIten
        i={0}
        todo={todosDemo[0]}
        handleThrough={handleThrough}
        handleDeleteIten={handleDeleteIten}
      />
    );

    wrapper.find("button").simulate("click");

    expect(handleDeleteIten).toHaveBeenCalledWith(todosDemo[0].id);
  });

  test("debe de mostrar el texto correctamente ", () => {
    const wrapper = shallow(
      <TodoListIten
        i={0}
        todo={todosDemo[0]}
        handleThrough={handleThrough}
        handleDeleteIten={handleDeleteIten}
      />
    );

    expect(wrapper.find("p").text().trim()).toBe(
      `${0 + 1}-${todosDemo[0].desc}`
    );
  });

  test("debe de mostrar el texto correctamente ", () => {
    const t = {
      id: 3,
      desc: "description3",
      done: true,
    };
    const wrapper = shallow(
      <TodoListIten
        i={0}
        todo={t}
        handleThrough={handleThrough}
        handleDeleteIten={handleDeleteIten}
      />
    );

    expect(wrapper.find("p").hasClass("complete")).toBe(true);
  });
});
