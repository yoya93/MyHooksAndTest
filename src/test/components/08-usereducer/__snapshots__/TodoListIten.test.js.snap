// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el TodoListIten debe mostrarse correctamente 1`] = `
<Fragment>
  <li
    className="list-group-iten mb-2"
    key="1"
  >
    <p
      className="false"
      onClick={[Function]}
    >
      1
      -
      description1
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Borrar
    </button>
  </li>
</Fragment>
`;
